src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,175 @@
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T replacedValue;
+        checkIndex(index);
+        Node current = getNodeByIndex(index);
+        replacedValue = (T) current.value;
+        current.value = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlinkNode(index);
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object
+                    || currentNode.value.equals(object) && currentNode.value != null) {
+                unlinkNode(i);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addAsHead(T value) {
+        if (size == 0) {
+            add(value);
+            return;
+        }
+        Node newNode = new Node(value);
+        newNode.next = head;
+        newNode.next.previous = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void insertBefore(Node currentNode, Node newNode) {
+        currentNode.previous.next = newNode;
+        newNode.previous = currentNode.previous;
+        currentNode.previous = newNode;
+        newNode.next = currentNode;
+        size++;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+    }
+
+    private Node getNodeByIndex(int index) {
+        Node currentNode;
+        if (size / 2 > index) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode;
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.previous;
+            }
+            return currentNode;
+        }
+    }
+
+    private T unlinkNode(int index) {
+        if (index == 0) {
+            return unlinkFirst(head);
+        }
+        checkIndex(index);
+        if (index == size - 1) {
+            return unlinkLast(tail);
+        }
+        Node unlinkedNode = getNodeByIndex(index);
+        unlinkedNode.previous.next = unlinkedNode.next;
+        unlinkedNode.next.previous = unlinkedNode.previous;
+        size--;
+        return (T) unlinkedNode.value;
+    }
+
+    private T unlinkFirst(Node node) {
```

Comment:
```suggestion
    private T unlinkFirst(Node<T> node) {
```
raw usage of a parametrized class is not suggested